File: backend/src/models/VirtualAccount.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IVirtualAccount extends Document {
  user: mongoose.Types.ObjectId;
  accountNumber: string;
  accountName: string;
  bankName: string;
  provider: string;
  reference: string;
  status: string;
  metadata: {
    virtualAccountName?: string;
    virtualAccountNo?: string;
    identityType?: string;
    licenseNumber?: string;
  };
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const VirtualAccountSchema = new Schema<IVirtualAccount>(
  {
    user: { 
      type: Schema.Types.ObjectId, 
      ref: 'User',
      required: true,
      index: true
    },
    accountNumber: { 
      type: String, 
      required: true,
      unique: true,
      index: true
    },
    accountName: { 
      type: String, 
      required: true 
    },
    bankName: { 
      type: String, 
      default: 'PalmPay',
      required: true 
    },
    provider: { 
      type: String, 
      default: 'payrant',
      enum: ['payrant', 'monnify', 'flutterwave'],
      required: true 
    },
    reference: { 
      type: String, 
      required: true,
      unique: true,
      index: true
    },
    status: { 
      type: String, 
      default: 'active',
      enum: ['active', 'inactive', 'suspended'],
      required: true 
    },
    metadata: {
      type: Schema.Types.Mixed,
      default: {}
    },
    isActive: { 
      type: Boolean, 
      default: true 
    }
  },
  {
    timestamps: true,
    toJSON: {
      transform: function(doc: any, ret: any) {
        ret.id = ret._id;
        delete ret._id;
        delete ret.__v;
        return ret;
      }
    }
  }
);

// Create a compound index for user and provider to ensure one account per provider per user
VirtualAccountSchema.index({ user: 1, provider: 1 }, { unique: true });

export default mongoose.model<IVirtualAccount>('VirtualAccount', VirtualAccountSchema);

File: backend-rust/src/models/virtual_account.rs
use serde::{Deserialize, Serialize};
use mongodb::bson::oid::ObjectId;
use chrono::{DateTime, Utc};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct VirtualAccount {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub user: ObjectId,
    pub account_number: String,
    pub account_name: String,
    pub bank_name: String,
    pub provider: String,
    pub reference: String,
    pub status: String,
    pub metadata: serde_json::Value,
    pub is_active: bool,
    pub generated_by: Option<ObjectId>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
